In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
from google.colab import drive
drive.mount('/content/drive')
import sys
# Check Python version
if sys.version_info.major == 3 and sys.version_info.minor >= 6:
  print("Hello, World!")
  print(f"Python version: {sys.version}")
else:
  print("Please use Python 3.6 or higher.")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Hello, World!
Python version: 3.11.11 (main, Dec  4 2024, 08:55:07) [GCC 11.4.0]


In [3]:
# Mutable Data Type Example (List)
my_list = [1, 2, 3]
print("Original List:", my_list)
# Modifying the list (changing contents)
my_list.append(4)
print("Modified List:", my_list)
# Immutable Data Type Example (Tuple)
my_tuple = (1, 2, 3)
print("Original Tuple:", my_tuple)

Original List: [1, 2, 3]
Modified List: [1, 2, 3, 4]
Original Tuple: (1, 2, 3)


In [4]:
age = 23 #int
pi = 3.14 #float
temperature = -5.5 #float
print("data type of variable age = ", type(age))
print("data type of variable pi = ", type(pi))
print("data type of variable temperature = ", type(temperature))

data type of variable age =  <class 'int'>
data type of variable pi =  <class 'float'>
data type of variable temperature =  <class 'float'>


In [5]:
name = "Alice"
greeting = 'Hello'
address = "123 Main St"
print("data type of the variable name = ",type(name))
print("data type of the variable greeting = ",type(greeting))
print("data type of the variable address = ",type(address))
# slice only one element
print("The first letter of the name is:", name[0])
print("The last letter of the name is:", name[-1])
# slice a range of elements
print("The second letter to the fourth of the name is:", name[1:4])
print("The first two letters of the name are:", name[:2])
print("Substring starting from the third letter is:", name[2:])

data type of the variable name =  <class 'str'>
data type of the variable greeting =  <class 'str'>
data type of the variable address =  <class 'str'>
The first letter of the name is: A
The last letter of the name is: e
The second letter to the fourth of the name is: lic
The first two letters of the name are: Al
Substring starting from the third letter is: ice


In [6]:
list1 = [1, 2, 3, 4]
mixed_list = [12, "Hello", True]
# List is mutable
mixed_list[0] = False
mixed_list

[False, 'Hello', True]

In [7]:
colors = ('red', 'green', 'yellow', 'blue')
print("First element:", colors[0])
print("Last two elements:", colors[2:])
print("Middle two elements:", colors[1:3])

First element: red
Last two elements: ('yellow', 'blue')
Middle two elements: ('green', 'yellow')


In [8]:
person = {'name':'John','age':30,'city':'Pittsburgh'}
print(f"Hello my name is {person['name']}. I am {person['age']} years old and I live at {person['city']}.")
print("All keys:", list(person.keys()))
print("All values:", list(person.values()))

Hello my name is John. I am 30 years old and I live at Pittsburgh.
All keys: ['name', 'age', 'city']
All values: ['John', 30, 'Pittsburgh']


In [9]:
unique_numbers = {1,2,3,3,3,3,4,5}
print(unique_numbers)

{1, 2, 3, 4, 5}


In [10]:
is_student = True
has_license = False
print("data type of the variable is_student = ",type(is_student))
print("data type of the variable has_license = ",type(has_license))

data type of the variable is_student =  <class 'bool'>
data type of the variable has_license =  <class 'bool'>


# To-do task

4.1 Exercise on Functions:

# Task - 1:
Create a Python program that converts between different units of measurement.
• The program should:
1. Prompt the user to choose the type of conversion (e.g., length, weight, volume).
2. Ask the user to input the value to be converted.
3. Perform the conversion and display the result.
4. Handle potential errors, such as invalid input or unsupported conversion types.

• Requirements:
1. Functions: Define at least one function to perform the conversion.
2. Error Handling: Use try-except blocks to handle invalid input (e.g., non-numeric values).
3. User Input: Prompt the user to select the conversion type and input the value.
4. Docstrings: Include a docstring in your function to describe its purpose, parameters, and
return value.

• Conversion Options:
1. Length:
– Convert meters (m) to feet (ft).
– Convert feet (ft) to meters (m).
2. Weight:
– Convert kilograms (kg) to pounds (lbs).
– Convert pounds (lbs) to kilograms (kg).
3. Volume:
– Convert liters (L) to gallons (gal).
– Convert gallons (gal) to liters (L).

In [11]:
def Length(unit,value):
  if unit=="m":
    return value * 3.28084
  elif unit=="ft":
    return value * 0.3048
  else:
    raise ValueError("Invalid unit. Use 'm' or 'ft'.")

def Weight(unit,value):
  if unit=="kg":
    return value * 2.20462
  elif unit=="lbs":
    return value * 0.453592
  else:
    raise ValueError("Invalid unit. Use 'kg' or 'lbs'.")

def Volume(unit,value):
  if unit=="L":
    return value * 0.26417
  elif unit=="gal":
    return value * 3.78541
  else:
    raise ValueError("Invalid unit. Use 'L' or 'gal'.")


def main():
  print("Choose a type of conversion: \n 1.Length(m<->ft) \n 2.Weight(kg<->lbs) \n 3.Volume(L<->gal)")
  try:
    choice = int(input("Enter your choice(1-3): "));
    if choice==1:
      unit= input("Enter unit (m/ft): ").lower()
      value= float(input("Enter value: "))
      result= Length(unit,value)
      converted_unit="ft" if unit =="m" else "m"
    elif choice==2:
      unit= input("Enter unit (kg/lbs): ").lower()
      value= float(input("Enter value: "))
      result= Weight(unit,value)
      converted_unit="lbs" if unit =="kg" else "kg"
    elif choice==3:
      unit= input("Enter unit (L/gal): ").lower()
      value= float(input("Enter value: "))
      result= Volume(unit,value)
      converted_unit="gal" if unit =="L" else "L"
    else:
      print("Invalid choice")
      return

    print(f"Converted Value: {result:.2f} {converted_unit}")
  except ValueError:
    print("Invalid input. Please enter a valid number.")
  except Exception as e:
    print(f"An error occurred: {e}")

if __name__=="__main__":
  main()

Choose a type of conversion: 
 1.Length(m<->ft) 
 2.Weight(kg<->lbs) 
 3.Volume(L<->gal)
Enter your choice(1-3): 1
Enter unit (m/ft): m
Enter value: 150
Converted Value: 492.13 ft


# Task - 2:
Create a Python program that performs various mathematical operations on a list of numbers.
• The Program should:
1. Prompt the user to choose an operation (e.g., find the sum, average, maximum, or minimum
of the numbers).
2. Ask the user to input a list of numbers (separated by spaces).
3. Perform the selected operation and display the result.
4. Handle potential errors, such as invalid input or empty lists.

• Requirements:
1. Functions: Define at least one function for each operation (sum, average, maximum, mini-
mum).
2. Error Handling: Use try-except blocks to handle invalid input (e.g., non-numeric values or
empty lists).
3. User Input: Prompt the user to select the operation and input the list of numbers.
4. Docstrings: Include a docstring in each function to describe its purpose, parameters, and
return value.

In [12]:
def calculate_sum(numbers):
    print(f"Calculating sum for: {numbers}")
    return sum(numbers)

def calculate_average(numbers):
    print(f"Calculating average for: {numbers}")
    return sum(numbers) / len(numbers) if numbers else 0

def find_maximum(numbers):
    print(f"Finding maximum for: {numbers}")
    return max(numbers)

def find_minimum(numbers):
    print(f"Finding minimum for: {numbers}")
    return min(numbers)


def main():
    print("Choose an operation: \n1.Sum \n2.Average \n3.Maximum \n4.Minimum")
    try:
        choice = int(input("Enter your choice(1-4): "))
        if choice not in [1, 2, 3, 4]:
            print("Invalid choice. Number should be between 1 and 4.")
            return

        numbers = input("Enter a list of numbers (separated by spaces): ").split()
        if not numbers:  # Handle case where input is empty
            print("Invalid input. Please enter a non-empty list of numbers.")
            return

        numbers = [float(num) for num in numbers]

        if choice == 1:
            result = calculate_sum(numbers)
            print(f"Sum: {result}")
        elif choice == 2:
            result = calculate_average(numbers)
            print(f"Average: {result}")
        elif choice == 3:
            result = find_maximum(numbers)
            print(f"Maximum: {result}")
        elif choice == 4:
            result = find_minimum(numbers)
            print(f"Minimum: {result}")

    except ValueError:
        print("Invalid input. Please enter numeric values only.")
    except Exception as e:
        print(f"An error occurred: {e}")

if __name__ == "__main__":
    main()


Choose an operation: 
1.Sum 
2.Average 
3.Maximum 
4.Minimum
Enter your choice(1-4): 1
Enter a list of numbers (separated by spaces): 5 4
Calculating sum for: [5.0, 4.0]
Sum: 9.0


# 4.2 Exercise on List Manipulation:

1. Extract Every Other Element:

Write a Python function that extracts every other element from a list, starting from the first element.

In [13]:
def extract_every_other(lst):
  return lst[::2]

print(extract_every_other([1,2,3,4,5,6]))

[1, 3, 5]


2. Slice a Sublist:

Write a Python function that returns a sublist from a given list, starting from a specified index and
ending at another specified index.

In [14]:
def get_sublist(lst, start, end):
  return lst[start:end+1]

print(get_sublist([1,2,3,4,5,6],2,4))

[3, 4, 5]


3. Reverse a List Using Slicing:

Write a Python function that reverses a list using slicing.

In [15]:
def reverse_list(lst):
  return lst[::-1];

print(reverse_list([1,2,3,4,5]))

[5, 4, 3, 2, 1]


4. Remove the First and Last Elements:

Write a Python function that removes the first and last elements of a list and returns the resulting
sublist.

In [16]:
def remove_first_last(lst):
  return lst[1:-1]

print(remove_first_last([1,2,3,4,5]))

[2, 3, 4]


5. Get the First n Elements:

Write a Python function that extracts the first n elements from a list.

In [17]:
def get_first_n(lst,n):
  return lst[0:n]

print(get_first_n([1,2,3,4,5],3))

[1, 2, 3]


6. Extract Elements from the End:

Write a Python function that extracts the last n elements of a list using slicing.

In [18]:
def get_last_n(lst,n):
  return lst[-n:]

print(get_last_n([1,2,3,4,5],3))

[3, 4, 5]


7. Extract Elements in Reverse Order:

Write a Python function that extracts a list of elements in reverse order starting from the second-to-last
element and skipping one element in between.

In [19]:
def reverse_skip(lst):
  return lst[-2::-2]

print(reverse_skip([1,2,3,4,5,6]))

[5, 3, 1]


# Exercise on Nested List:

1. Flatten a Nested List:

Write a Python function that takes a nested list and flattens it into a single list, where all the elements
are in a single dimension.

In [20]:
def flatten(lst):
  flat_list=[]
  for sublist in lst:
    flat_list.extend(sublist)
  return flat_list

print(flatten([[1,2],[3,4],[5]]))

[1, 2, 3, 4, 5]


2. Accessing Nested List Elements:

Write a Python function that extracts a specific element from a nested list given its indices.

In [21]:
def access_nested_element(lst, indices):
  return lst[indices[0]][indices[1]]

print(access_nested_element([[1,2,3],[4,5,6],[7,8,9]],[1,2]))

6


3. Sum of All Elements in a Nested List:

Write a Python function that calculates the sum of all the numbers in a nested list (regardless of depth).

In [22]:
def sum_nested(lst):
    total = 0

    for element in lst:
        if isinstance(element, list):
            total += sum_nested(element)
        else:
            total += element
    return total
print(sum_nested([[1, 2], [3, [4, 5]], 6]))

21


4. Remove Specific Element from a Nested List:

Write a Python function that removes all occurrences of a specific element from a nested list.

In [23]:
def remove_element(lst, elem):
  result=[]

  for item in lst:
    if isinstance(item, list):
      filtered_sublist= remove_element(item, elem)
      if filtered_sublist:
        result.append(filtered_sublist)
    elif item != elem:
      result.append(item)
  return result

print(remove_element([[1,2],[3,2],[4,5]],2))

[[1], [3], [4, 5]]


5. Find the Maximum Element in a Nested List:

Write a Python function that finds the maximum element in a nested list (regardless of depth).


In [24]:
def find_max(lst):
    max_value = float('-inf')
    for item in lst:
        if isinstance(item, list):
            max_value = max(max_value, find_max(item))
        else:
            max_value = max(max_value, item)

    return max_value

nested_list = [[1, 2], [3, [4, 5]], 6]
print(find_max(nested_list))

6


6. Count Occurrences of an Element in a Nested List:

Write a Python function that counts how many times a specific element appears in a nested list.

In [25]:
def count_occurrences(lst, elem):
    count = 0

    for item in lst:
        if isinstance(item, list):
            count += count_occurrences(item, elem)
        elif item == elem:
            count += 1

    return count
lst = [[1, 2], [2, 3], [2, 4]]
elem = 2
print(f"Occurrences of {elem}:", count_occurrences(lst, elem))

Occurrences of 2: 3


7. Flatten a List of Lists of Lists:

Write a Python function that flattens a list of lists of lists into a single list, regardless of the depth.

In [26]:
def deep_flatten(lst):
    flattened_list = []

    for item in lst:
        if isinstance(item, list):
          flattened_list.extend(deep_flatten(item))
        else:
            flattened_list.append(item)

    return flattened_list
nested_list = [[[1, 2], [3, 4]], [[5, 6], [7, 8]]]
print(deep_flatten(nested_list))

[1, 2, 3, 4, 5, 6, 7, 8]


8. Nested List Average:

Write a Python function that calculates the average of all elements in a nested list.

In [27]:
def average_nested(lst):
    total_sum = 0
    total_count = 0
    def helper(sublist):
        nonlocal total_sum, total_count
        for item in sublist:
            if isinstance(item, list):
                helper(item)
            else:
                total_sum += item
                total_count += 1
    helper(lst)
    return total_sum / total_count if total_count > 0 else 0
nested_list = [[1, 2], [3, 4], [5, 6]]
print(average_nested(nested_list))


3.5


# About a Numpy.

In [28]:
import numpy as np
# initialize an all zero array with size 2 X 3:
zeros_arr = np.zeros((2,3))
print(" A zeros array is \n", zeros_arr, "with dimensions", zeros_arr.shape, "\n")

 A zeros array is 
 [[0. 0. 0.]
 [0. 0. 0.]] with dimensions (2, 3) 



In [29]:
import numpy as np
# initialize an all zero array with size 2 X 3:
zeros_arr = np.zeros((2,3))
print(" A zeros array is \n", zeros_arr, "with dimensions", zeros_arr.shape, "\n")

 A zeros array is 
 [[0. 0. 0.]
 [0. 0. 0.]] with dimensions (2, 3) 



In [30]:
import numpy as np
# Create and display zero, one, and n-dimensional arrays
zero_dim_array = np.array(5)
one_dim_array = np.array([1, 2, 3])
n_dim_array = np.array([[1, 2], [3, 4]])
for arr in [zero_dim_array, one_dim_array, n_dim_array]:
  print(f"Array:\n{arr}\nDimension: {arr.ndim}\nData type: {arr.dtype}\n")

Array:
5
Dimension: 0
Data type: int64

Array:
[1 2 3]
Dimension: 1
Data type: int64

Array:
[[1 2]
 [3 4]]
Dimension: 2
Data type: int64



In [31]:
# For a 1-D array:
import numpy as np
arr_1d = np.array([1, 2, 3, 4, 5])
print(arr_1d.ndim) # Output: 1
# For a 2-D array:
arr_2d = np.array([[1, 2, 3],
[4, 5, 6]])
print(arr_2d.ndim) # Output: 2
#For a 3-D array:
arr_3d = np.array([[[1, 2, 3],
[4, 5, 6]],
[[7, 8, 9],
[10, 11, 12]]])
print(arr_3d.ndim) # Output: 3

1
2
3


In [32]:
print(arr_1d.shape) # Output: (5,)
print(arr_2d.shape) # Output: (2, 3)
print(arr_3d.shape) # Output: (2, 2, 3)

(5,)
(2, 3)
(2, 2, 3)


In [33]:
print(arr_1d.size) # Output: 5
print(arr_2d.size) # Output: 6
print(arr_3d.size) # Output: 12

5
6
12


In [34]:
print(arr_1d.dtype) # Output: int64 (or int32)
arr_float = np.array([1.1, 2.2, 3.3])
print(arr_float.dtype) # Output: float64

int64
float64


In [35]:
print(arr_1d.itemsize) # Output: 8 (for int64)

8


In [36]:
print(arr_1d.nbytes) # Output: 40 (5 elements \times 8 bytes each for int64)

40


In [37]:
print(arr_2d.T)


[[1 4]
 [2 5]
 [3 6]]


In [38]:
for item in arr_2d.flat:
  print(item, end=" ") # Output: 1 2 3 4 5 6

1 2 3 4 5 6 

In [39]:
arr_complex = np.array([1+2j, 3+4j])
print(arr_complex.real) # Output: [1. 3.]
print(arr_complex.imag) # Output: [2. 4.]

[1. 3.]
[2. 4.]


In [40]:
arr_float = arr_2d.astype(float)
print(arr_float)
# Output:
# [[1. 2. 3.]
# [4. 5. 6.]]

[[1. 2. 3.]
 [4. 5. 6.]]


In [41]:
lst = [[1, 2], [3, 4]]
# print(lst.shape)
# Error: List has no .shape attribute
#Fixing the Error:
arr = np.array(lst)
print(arr.shape) # Output: (2, 2)

(2, 2)


In [42]:
arr = np.array([1, 2, 3])
# print(arr.shape[1]) # Error: 1D array only has one dimension
# Fixing the Error
if len(arr.shape) > 1:
  print(arr.shape[1]) # Access only if it’s valid

In [43]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
# print(arr.shape()) # Error: .shape is an attribute, not a function
print(arr.shape) # Correct usage

(2, 3)


In [44]:
# Create an array with shape (2, 3)
array = np.array([[1, 2, 3],
[4, 5, 6]])

# Reshape to (3, 2), keeping 6 elements
reshaped_array = array.reshape(3, 2)
print("Original Shape:", array.shape, "\nReshaped Shape:", reshaped_array.shape)
# Expected Outputs:
# Original Shape: (2, 3)

Original Shape: (2, 3) 
Reshaped Shape: (3, 2)


In [45]:
import numpy as np
a = np.arange(1, 10)
print(a)
x = range(1, 10)
print(x) # x is an iterator
print(list(x))
# further arange examples:
x = np.arange(10.4)
print(x)
x = np.arange(0.5, 10.4, 0.8)
print(x)

[1 2 3 4 5 6 7 8 9]
range(1, 10)
[1, 2, 3, 4, 5, 6, 7, 8, 9]
[ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10.]
[ 0.5  1.3  2.1  2.9  3.7  4.5  5.3  6.1  6.9  7.7  8.5  9.3 10.1]


In [46]:
import numpy as np
# Create arrays of specified shapes
ones_array = np.ones((2, 3)) # Shape: (2, 3)
zeros_array = np.zeros((3, 2)) # Shape: (3, 2)
empty_array = np.empty((2, 2)) # Shape: (2, 2)
identity_matrix = np.eye(3) # Shape: (3, 3)
print(ones_array, zeros_array, empty_array, identity_matrix, sep='\n\n')

[[1. 1. 1.]
 [1. 1. 1.]]

[[0. 0.]
 [0. 0.]
 [0. 0.]]

[[4.9e-324 9.9e-324]
 [1.5e-323 2.0e-323]]

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [47]:
import numpy as np
array_from_list = np.array([1, 2, 3]) # [1 2 3]
array_from_tuple = np.array((4, 5, 6)) # [4 5 6]
array_from_nested_list = np.array([[1, 2, 3], [4, 5, 6]]) # [[1 2 3] [4 5 6]]
print(array_from_list, array_from_tuple, array_from_nested_list, sep='\n')

[1 2 3]
[4 5 6]
[[1 2 3]
 [4 5 6]]


In [48]:
import numpy as np
# Existing Array of Shape:(2,3)
arr = np.array([[1, 2, 3], [4, 5, 6]])
# Creating Array with Shape of existing array:
zeros, ones, empty = np.zeros(arr.shape), np.ones(arr.shape), np.empty(arr.shape) #
Shape: (2,3)
print(arr, zeros, ones, empty, sep='\n')

[[1 2 3]
 [4 5 6]]
[[0. 0. 0.]
 [0. 0. 0.]]
[[1. 1. 1.]
 [1. 1. 1.]]
[[2.47227954e-316 0.00000000e+000 0.00000000e+000]
 [8.95593996e-312 2.10966031e-321 4.23754071e-307]]


In [49]:
import numpy as np
arr1 = np.array([[1, 2], [3, 4]]) # Shape: (2, 2)
arr2 = np.array([[5, 6], [7, 8]]) # Shape: (2, 2)
# Stacking examples
hstacked = np.hstack((arr1, arr2))
# Output hstack: [[1 2 5 6] [3 4 7 8]]
vstacked = np.vstack((arr1, arr2))
# Output vstack[[1 2] [3 4] [5 6] [7 8]]
colstacked = np.column_stack((arr1[0], arr2[0]))
# Output column stack: [[1 5] [2 6]]
print(hstacked, vstacked, colstacked, sep='\n')

[[1 2 5 6]
 [3 4 7 8]]
[[1 2]
 [3 4]
 [5 6]
 [7 8]]
[[1 5]
 [2 6]]


In [50]:
import numpy as np
# Creating two 3D arrays
arr1 = np.array([[[1], [2]], [[3], [4]]]) # Shape: (2, 2, 1)
arr2 = np.array([[[5], [6]], [[7], [8]]]) # Shape: (2, 2, 1)
# Concatenate along axis 0 (vertical stacking)
concat_axis0 = np.concatenate((arr1, arr2), axis=0)
# Shape of concat_axis0: (4, 2, 1), arr1 is stacked on top of arr2
# Concatenate along axis 1 (horizontal stacking)
concat_axis1 = np.concatenate((arr1, arr2), axis=1)
# Shape of concat_axis1: (2, 4, 1), arr1 and arr2 are placed side by side
# Concatenate along axis 2 (depth stacking)
concat_axis2 = np.concatenate((arr1, arr2), axis=2)
# Shape of concat_axis2: (2, 2, 2), arr1 and arr2 are stacked in depth (third dimension)
# Displaying the results
print("Concatenation along axis 0 (Vertical stacking):\n", concat_axis0)
print("Shape of concatenated array along axis 0:", concat_axis0.shape)
print("\nConcatenation along axis 1 (Horizontal stacking):\n", concat_axis1)
print("Shape of concatenated array along axis 1:", concat_axis1.shape)
print("\nConcatenation along axis 2 (Depth stacking):\n", concat_axis2)
print("Shape of concatenated array along axis 2:", concat_axis2.shape)

Concatenation along axis 0 (Vertical stacking):
 [[[1]
  [2]]

 [[3]
  [4]]

 [[5]
  [6]]

 [[7]
  [8]]]
Shape of concatenated array along axis 0: (4, 2, 1)

Concatenation along axis 1 (Horizontal stacking):
 [[[1]
  [2]
  [5]
  [6]]

 [[3]
  [4]
  [7]
  [8]]]
Shape of concatenated array along axis 1: (2, 4, 1)

Concatenation along axis 2 (Depth stacking):
 [[[1 5]
  [2 6]]

 [[3 7]
  [4 8]]]
Shape of concatenated array along axis 2: (2, 2, 2)


In [51]:
import numpy as np
# Incorrect shapes
arr1 = np.array([[1, 2], [3, 4]]) # shape (2, 2)
arr2 = np.array([5, 6]) # shape (2,)
# Attempting horizontal stack
try:
  hstacked = np.hstack((arr1, arr2)) # Error due to dimension mismatch
except ValueError as e:
  print(f"Error: {e}")
# Solution: Reshape arr2 to match arr1’s dimensions
arr2 = np.array([5, 6]).reshape(2, 1) # shape (2, 1)
hstacked = np.hstack((arr1, arr2)) # No error
print(hstacked)

Error: all the input arrays must have same number of dimensions, but the array at index 0 has 2 dimension(s) and the array at index 1 has 1 dimension(s)
[[1 2 5]
 [3 4 6]]


In [52]:
import numpy as np
arr1 = np.array([[1, 2, 3], [4, 5, 6]]) # shape (2, 3)
arr2 = np.array([[7, 8], [9, 10]]) # shape (2, 2)
# Attempting horizontal stack (this will fail)
try:
  hstacked = np.hstack((arr1, arr2)) # Error due to shape mismatch
except ValueError as e:
  print(f"Error: {e}")
# Solution: Make sure the number of rows in both arrays is the same before stacking
arr2 = np.array([[7, 8, 9], [10, 11, 12]]) # shape (2, 3)
hstacked = np.hstack((arr1, arr2)) # Works fine
print(hstacked)

[[ 1  2  3  7  8  9]
 [ 4  5  6 10 11 12]]


In [53]:
arr1 = np.array([[[1, 2], [3, 4]]]) # shape (1, 2, 2)
arr2 = np.array([[5, 6], [7, 8]]) # shape (2, 2)
# Attempting stacking with 3D and 2D arrays
try:
  vstacked = np.vstack((arr1, arr2)) # Error due to too many dimensions
except ValueError as e:
  print(f"Error: {e}")
# Solution: Reshape arr1 to be 2D (or flatten it) before stacking
arr1 = arr1.reshape(2, 2) # Now it’s shape (2, 2)
vstacked = np.vstack((arr1, arr2)) # Works fine
print(vstacked)

Error: all the input arrays must have same number of dimensions, but the array at index 0 has 3 dimension(s) and the array at index 1 has 2 dimension(s)
[[1 2]
 [3 4]
 [5 6]
 [7 8]]


In [54]:
import numpy as np
# Creating a 2D array
array = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
# Accessing elements using indexing
print(array[0, 1]) # Output: 2 (first row, second column)
print(array[-1, -1]) # Output: 9 (last row, last column)
# Slicing the array
print(array[0:2, 1:3]) # Output: [[2, 3], [5, 6]]

2
9
[[2 3]
 [5 6]]


In [55]:
import numpy as np
# Creating a 2D array
array = np.array([[1, 2, 3], [4, 5, 6]])
# Attempting to access an element out of bounds
try:
  print(array[2, 1])
except IndexError as e:
  print(f"Error: {e}")

Error: index 2 is out of bounds for axis 0 with size 2


In [56]:
import numpy as np
# Example of basic ufuncs: square and square root
arr = np.array([1, 4, 9, 16])
# Square every element
squared = np.square(arr)
# Square root of every element
sqrt = np.sqrt(arr)
print(f"Squared: {squared}")
print(f"Square root: {sqrt}")

Squared: [  1  16  81 256]
Square root: [1. 2. 3. 4.]


In [57]:
import numpy as np
# Trigonometric ufuncs
angles = np.array([0, np.pi / 4, np.pi / 2])
sine_values = np.sin(angles)
# Logarithmic ufuncs
log_values = np.log(arr)

In [58]:
import numpy as np
# Matrix multiplication
A = np.array([[1, 2], [3, 4]])
B = np.array([[5, 6], [7, 8]])
product = np.dot(A, B)
# Compute determinant
det = np.linalg.det(A)
# Compute inverse
inverse = np.linalg.inv(A)
# Compute eigenvalues and eigenvectors
eigvals, eigvecs = np.linalg.eig(A)
print(f"Matrix Product:\n{product}")
print(f"Determinant: {det}")
print(f"Inverse:\n{inverse}")
print(f"Eigenvalues: {eigvals}, Eigenvectors:\n{eigvecs}")

Matrix Product:
[[19 22]
 [43 50]]
Determinant: -2.0000000000000004
Inverse:
[[-2.   1. ]
 [ 1.5 -0.5]]
Eigenvalues: [-0.37228132  5.37228132], Eigenvectors:
[[-0.82456484 -0.41597356]
 [ 0.56576746 -0.90937671]]


In [59]:
import numpy as np
import time
# Create two large arrays
size = 10**6
a = np.random.rand(size)
b = np.random.rand(size)
# Using a loop
start_time = time.time()
result_loop = np.zeros(size)
for i in range(size):
  result_loop[i] = a[i] + b[i]
end_time = time.time()
loop_time = end_time - start_time
print(f"Loop Execution Time: {loop_time:.5f} seconds")
# Using NumPy’s vectorized operations
start_time = time.time()
result_vectorized = a + b # Element-wise addition
end_time = time.time()
vectorized_time = end_time - start_time
print(f"Vectorized Execution Time: {vectorized_time:.5f} seconds")
# Performance improvement factor
speedup = loop_time / vectorized_time
print(f"Vectorization is approximately {speedup:.2f} times faster!")

Loop Execution Time: 0.93132 seconds
Vectorized Execution Time: 0.00525 seconds
Vectorization is approximately 177.38 times faster!


# To - Do - NumPy

10.1 Basic Vector and Matrix Operation with Numpy.

Problem - 1: Array Creation:

In [60]:
import numpy as np

# 1. Initialize an empty array with size 2x2
empty_array = np.empty((2, 2))
print("Empty Array (2x2):\n", empty_array)

# 2. Initialize an all-ones array with size 4x2
ones_array = np.ones((4, 2))
print("\nAll Ones Array (4x2):\n", ones_array)

# 3. Return a new array of given shape and type, filled with fill value
fill_value = 7
filled_array = np.full((3, 3), fill_value)
print("\nArray filled with", fill_value, ":\n", filled_array)

# 4. Return a new array of zeros with same shape and type as a given array
sample_array = np.array([[1, 2, 3], [4, 5, 6]])
zeros_like_array = np.zeros_like(sample_array)
print("\nZeros-like Array:\n", zeros_like_array)

# 5. Return a new array of ones with same shape and type as a given array
ones_like_array = np.ones_like(sample_array)
print("\nOnes-like Array:\n", ones_like_array)

# 6. Convert an existing list to a numpy array
new_list = [1, 2, 3, 4]
numpy_array = np.array(new_list)
print("\nConverted NumPy Array:\n", numpy_array)

Empty Array (2x2):
 [[0.82456484 0.41597356]
 [0.56576746 0.90937671]]

All Ones Array (4x2):
 [[1. 1.]
 [1. 1.]
 [1. 1.]
 [1. 1.]]

Array filled with 7 :
 [[7 7 7]
 [7 7 7]
 [7 7 7]]

Zeros-like Array:
 [[0 0 0]
 [0 0 0]]

Ones-like Array:
 [[1 1 1]
 [1 1 1]]

Converted NumPy Array:
 [1 2 3 4]


Problem - 2: Array Manipulation: Numerical Ranges and Array indexing:

In [61]:
import numpy as np

# 1. Create an array with values ranging from 10 to 49
array_range = np.arange(10, 50)
print("Array with values from 10 to 49:\n", array_range)

# 2. Create a 3X3 matrix with values ranging from 0 to 8
matrix_3x3 = np.arange(9).reshape(3, 3)
print("\n3X3 Matrix with values from 0 to 8:\n", matrix_3x3)

# 3. Create a 3X3 identity matrix
identity_matrix = np.eye(3)
print("\n3X3 Identity Matrix:\n", identity_matrix)

# 4. Create a random array of size 30 and find the mean of the array
random_array = np.random.random(30)
mean_value = random_array.mean()
print("\nMean of random array:\n", mean_value)

# 5. Create a 10X10 array with random values and find the min & max values
random_matrix = np.random.random((10, 10))
min_value = random_matrix.min()
max_value = random_matrix.max()
print("\nMinimum value in 10X10 matrix:", min_value)
print("Maximum value in 10X10 matrix:", max_value)

# 6. Create a zero array of size 10 and replace 5th element with 1
zero_array = np.zeros(10)
zero_array[4] = 1  # Index 4 is the 5th element (0-based index)
print("\nZero array with 5th element replaced by 1:\n", zero_array)

# 7. Reverse an array arr = [1,2,0,0,4,0]
arr = np.array([1, 2, 0, 0, 4, 0])
reversed_arr = arr[::-1]
print("\nReversed array:\n", reversed_arr)

# 8. Create a 2D array with 1 on the border and 0 inside
border_array = np.ones((5, 5))
border_array[1:-1, 1:-1] = 0
print("\n2D array with 1 on border and 0 inside:\n", border_array)

# 9. Create an 8X8 matrix and fill it with a checkerboard pattern
checkerboard = np.zeros((8, 8), dtype=int)
checkerboard[1::2, ::2] = 1
checkerboard[::2, 1::2] = 1
print("\n8X8 Checkerboard Pattern:\n", checkerboard)


Array with values from 10 to 49:
 [10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33
 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49]

3X3 Matrix with values from 0 to 8:
 [[0 1 2]
 [3 4 5]
 [6 7 8]]

3X3 Identity Matrix:
 [[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]

Mean of random array:
 0.44127680823190324

Minimum value in 10X10 matrix: 0.004247053758072661
Maximum value in 10X10 matrix: 0.993214493934577

Zero array with 5th element replaced by 1:
 [0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]

Reversed array:
 [0 4 0 0 2 1]

2D array with 1 on border and 0 inside:
 [[1. 1. 1. 1. 1.]
 [1. 0. 0. 0. 1.]
 [1. 0. 0. 0. 1.]
 [1. 0. 0. 0. 1.]
 [1. 1. 1. 1. 1.]]

8X8 Checkerboard Pattern:
 [[0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]]


Problem - 3: Array Operations:

In [62]:
import numpy as np

# Given arrays
x = np.array([[1, 2], [3, 5]])
y = np.array([[5, 6], [7, 8]])
v = np.array([9, 10])
w = np.array([11, 12])

# 1. Add the two arrays
addition = x + y
print("Addition of x and y:\n", addition)

# 2. Subtract the two arrays
subtraction = x - y
print("\nSubtraction of x and y:\n", subtraction)

# 3. Multiply the array x with an integer
multiplied_x = x * 3
print("\nMultiplication of x with 3:\n", multiplied_x)

# 4. Find the square of each element of array x
square_x = np.square(x)
print("\nSquare of each element in x:\n", square_x)

# 5. Find the dot product
dot_vw = np.dot(v, w)
dot_xv = np.dot(x, v)
dot_xy = np.dot(x, y)

print("\nDot product of v and w:", dot_vw)
print("Dot product of x and v:\n", dot_xv)
print("Dot product of x and y:\n", dot_xy)

# 6. Concatenate x and y along rows, and v and w along columns
concat_xy_row = np.concatenate((x, y), axis=0)
concat_vw_col = np.vstack((v, w))

print("\nConcatenation of x and y along rows:\n", concat_xy_row)
print("\nConcatenation of v and w along columns:\n", concat_vw_col)

# 7. Attempt to concatenate x and v
try:
    concat_xv = np.concatenate((x, v), axis=0)
    print("\nConcatenation of x and v:\n", concat_xv)
except ValueError as e:
    print("\nError while concatenating x and v:", e)


Addition of x and y:
 [[ 6  8]
 [10 13]]

Subtraction of x and y:
 [[-4 -4]
 [-4 -3]]

Multiplication of x with 3:
 [[ 3  6]
 [ 9 15]]

Square of each element in x:
 [[ 1  4]
 [ 9 25]]

Dot product of v and w: 219
Dot product of x and v:
 [29 77]
Dot product of x and y:
 [[19 22]
 [50 58]]

Concatenation of x and y along rows:
 [[1 2]
 [3 5]
 [5 6]
 [7 8]]

Concatenation of v and w along columns:
 [[ 9 10]
 [11 12]]

Error while concatenating x and v: all the input arrays must have same number of dimensions, but the array at index 0 has 2 dimension(s) and the array at index 1 has 1 dimension(s)


x is a 2×2 matrix (shape = (2,2)). v is a 1D array (shape = (2,)). When concatenating along axis=0 (rows), both arrays must have the same number of columns. Since x has two columns but v is a 1D array with no explicit columns, NumPy throws a shape mismatch error.

Problem - 4: Matrix Operations:

In [63]:
# For the following arrays:
# A = np.array([[3,4],[7,8]]) and B = np.array([[5,3],[2,1]]);
# Prove following with Numpy:
A = np.array([[3, 4], [7, 8]])
B = np.array([[5, 3], [2, 1]])

In [64]:
# 1. Prove A.A−1 = I.
A_inv = np.linalg.inv(A)  # Find inverse of A
identity_matrix = np.dot(A, A_inv)  # Multiply A with A^-1
identity_matrix = np.round(identity_matrix, decimals=5)  # Round to avoid floating-point errors

print("A * A^-1:\n", identity_matrix)


A * A^-1:
 [[1. 0.]
 [0. 1.]]


In [65]:
# 2. Prove AB ̸= BA.
AB = np.dot(A, B)  # A * B
BA = np.dot(B, A)  # B * A

print("\nAB:\n", AB)
print("\nBA:\n", BA)

# Check if AB is equal to BA
is_commutative = np.array_equal(AB, BA)
print("\nIs AB equal to BA?", is_commutative)



AB:
 [[23 13]
 [51 29]]

BA:
 [[36 44]
 [13 16]]

Is AB equal to BA? False


In [66]:
# 3. Prove (AB)^T = B^T*A^T
AB_T = np.transpose(AB)  # (AB)^T
BT_AT = np.dot(np.transpose(B), np.transpose(A))  # B^T * A^T

print("\n(AB)^T:\n", AB_T)
print("\nB^T * A^T:\n", BT_AT)

# Check equality
is_transpose_property = np.array_equal(AB_T, BT_AT)
print("\nIs (AB)^T equal to B^T * A^T?", is_transpose_property)



(AB)^T:
 [[23 51]
 [13 29]]

B^T * A^T:
 [[23 51]
 [13 29]]

Is (AB)^T equal to B^T * A^T? True


In [67]:
# Solve the following system of Linear equation using Inverse Methods.

# 2x − 3y + z = −1
# x − y + 2z = −3
# 3x + y − z = 9


# Coefficient matrix A
A = np.array([
    [2, -3, 1],
    [1, -1, 2],
    [3, 1, -1]
])

# Constant matrix B
B = np.array([-1, -3, 9])
A_inv = np.linalg.inv(A)  # Find inverse of A
X = np.dot(A_inv, B)  # Compute X = A^-1 * B

print("\nSolution for x, y, z:\n", X)
X = np.linalg.solve(A, B)
print("\nSolution using np.linalg.solve:\n", X)



Solution for x, y, z:
 [ 2.  1. -2.]

Solution using np.linalg.solve:
 [ 2.  1. -2.]


10.2 Experiment: How Fast is Numpy?

1. Element-wise Addition:

In [68]:
# • Using Python Lists, perform element-wise addition of two lists of size 1, 000, 000. Measure and Print the time taken for this operation.
import time
import random
size = 1_000_000
list1 = [random.randint(1, 100) for _ in range(size)]
list2 = [random.randint(1, 100) for _ in range(size)]
start_time = time.time()
result = [list1[i] + list2[i] for i in range(size)]
end_time = time.time()
print(f"Time taken for element-wise addition of lists: {end_time - start_time:.5f} seconds")


Time taken for element-wise addition of lists: 0.05683 seconds


In [69]:
# Using Numpy Arrays, Repeat the calculation and measure and print the time taken for this operation.
import numpy as np
arr1 = np.random.randint(1, 100, size)
arr2 = np.random.randint(1, 100, size)
start_time = time.time()
result_np = arr1 + arr2
end_time = time.time()
print(f"Time taken using NumPy: {end_time - start_time:.5f} seconds")

Time taken using NumPy: 0.00470 seconds


2. Element-wise Multiplication

In [70]:
import time
import random
import numpy as np
size = 1_000_000
list1 = [random.randint(1, 100) for _ in range(size)]
list2 = [random.randint(1, 100) for _ in range(size)]
start_time_list_mul = time.time()
result_list_mul = [list1[i] * list2[i] for i in range(size)]
end_time_list_mul = time.time()
time_list_mul = end_time_list_mul - start_time_list_mul
print(f"Time taken for element-wise multiplication using Python lists: {time_list_mul:.5f} seconds")

arr1 = np.random.randint(1, 100, size)
arr2 = np.random.randint(1, 100, size)
start_time_np_mul = time.time()
result_np_mul = arr1 * arr2
end_time_np_mul = time.time()
time_np_mul = end_time_np_mul - start_time_np_mul

print(f"Time taken for element-wise multiplication using NumPy arrays: {time_np_mul:.5f} seconds")


Time taken for element-wise multiplication using Python lists: 0.08511 seconds
Time taken for element-wise multiplication using NumPy arrays: 0.00509 seconds


3. Dot Product

In [71]:
import time
import random
import numpy as np
size = 1_000_000
list1 = [random.randint(1, 100) for _ in range(size)]
list2 = [random.randint(1, 100) for _ in range(size)]

start_time_list_dot = time.time()
dot_product_list = sum(list1[i] * list2[i] for i in range(size))
end_time_list_dot = time.time()
time_list_dot = end_time_list_dot - start_time_list_dot

print(f"Time taken for dot product using Python lists: {time_list_dot:.5f} seconds")

arr1 = np.random.randint(1, 100, size)
arr2 = np.random.randint(1, 100, size)

start_time_np_dot = time.time()
dot_product_np = np.dot(arr1, arr2)
end_time_np_dot = time.time()
time_np_dot = end_time_np_dot - start_time_np_dot

print(f"Time taken for dot product using NumPy arrays: {time_np_dot:.5f} seconds")

Time taken for dot product using Python lists: 0.08377 seconds
Time taken for dot product using NumPy arrays: 0.00179 seconds


4. Matrix Multiplication

In [72]:
import numpy as np
import time

matrix_size = 1000
matrix1_list = [[np.random.randint(1, 100) for _ in range(matrix_size)] for _ in range(matrix_size)]
matrix2_list = [[np.random.randint(1, 100) for _ in range(matrix_size)] for _ in range(matrix_size)]
start_time_list = time.time()
result_matrix_list = [[sum(matrix1_list[i][k] * matrix2_list[k][j] for k in range(matrix_size))
                       for j in range(matrix_size)] for i in range(matrix_size)]
end_time_list = time.time()
time_list_matmul = end_time_list - start_time_list
print(f"Time taken for matrix multiplication using Python lists: {time_list_matmul:.5f} seconds")

matrix1_np = np.random.randint(1, 100, (matrix_size, matrix_size))
matrix2_np = np.random.randint(1, 100, (matrix_size, matrix_size))
start_time_np = time.time()
result_matrix_np = np.dot(matrix1_np, matrix2_np)
end_time_np = time.time()
time_np_matmul = end_time_np - start_time_np
print(f"Time taken for matrix multiplication using NumPy: {time_np_matmul:.5f} seconds")


Time taken for matrix multiplication using Python lists: 131.55584 seconds
Time taken for matrix multiplication using NumPy: 1.50634 seconds


In [73]:
import random
import time

size = 1000
A = [[random.randint(1, 10) for _ in range(size)] for _ in range(size)]
B = [[random.randint(1, 10) for _ in range(size)] for _ in range(size)]

def matrix_multiply(A, B):
    result = [[0] * size for _ in range(size)]
    for i in range(size):
        for j in range(size):
            for k in range(size):
                result[i][j] += A[i][k] * B[k][j]
    return result
start_time = time.time()
C = matrix_multiply(A, B)
end_time = time.time()
print(f"Time taken using Python lists: {end_time - start_time:.5f} seconds")

Time taken using Python lists: 139.24335 seconds


In [74]:
import numpy as np

A_np = np.random.randint(1, 11, (size, size))
B_np = np.random.randint(1, 11, (size, size))

start_time = time.time()
C_np = np.dot(A_np, B_np)
end_time = time.time()

print(f"Time taken using NumPy: {end_time - start_time:.5f} seconds")


Time taken using NumPy: 1.61509 seconds
